File: Prácticas/P2/Ruby/civitas/lib/jugador.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "dado.rb"
require_relative "diario.rb"
require_relative "titulo_propiedad.rb"
require_relative "sorpresa.rb"

module Civitas
  class Jugador
    attr_reader :propiedades, :casas_max, :casas_por_hotel, :encarcelado, :hoteles_max, :nombre, :num_casilla_actual, :precio_libertad, :puede_comprar, :saldo, :paso_por_salida
    @@CASAS_MAX = 4
    @@CASAS_POR_HOTEL = 4
    @@HOTELES_MAX = 4
    @@PASO_POR_SALIDA = 1000
    @@PRECIO_LIBERTAD = 200
    @@SALDO_INICIAL = 7500
    
    def initialize(nombre)
      @nombre = nombre
      @encarcelado = false
      @propiedades = Array.new
      @saldo = @@SALDO_INICIAL
      @puede_comprar = false
      @diario = Diario.instance
      @dado = Dado.instance
      @num_casilla_actual = 0
      @salvoconducto = nil
    end
    
    def self.ini_otro(otro)
      self.new(otro.nombre)
      @encarcelado = otro.encarcelado
      @propiedades = otro.propiedades
      @saldo = otro.saldo
      @puede_comprar = otro.puede_comprar
      @num_casilla_actual = otro.num_casilla_actual
    end
    
    private
    
    def debe_ser_encarcelado
      salida = false
      if(!@encarcelado)
        if(!tiene_salvoconducto)
          salida = true
        else
          perder_salvoconducto()
          @diario.ocurre_evento("Se libra de la carcel el jugador: " + @nombre)
        end
      end
      salida
    end
    
    def perder_salvo_conducto()
      @salvoconducto.usadas()
      @salvoconducto = nil
    end
    
    def puede_salir_carcel_pagando()
      @saldo >= @@PRECIO_LIBERTAD
    end
    
    def puede_edificar_casa(propiedad)
      (puedo_gastar(propiedad.precio_edificar) && propiedad.num_casas < 4)
    end
    
    def puede_edificar_hotel(propiedad)
      (puedo_gastar(propiedad.precio_edificar) && propiedad.num_hoteles < 4 && propiedad.num_casas == 4)
    end
    
    def puedo_gastar(precio)
      !@encarcelado && @saldo >= precio
    end
    
    public
    
    def cancalar_hipoteca(ip)
      
    end
    
    def cantidad_casas_hoteles()
      suma = 0
      for i in 1..@propiedades.lenght
        suma = suma + @propiedades.at(i).num_casas + @propiedades.at(i).num_hoteles
      end
      suma
    end
    
    def <=>(otroJugador)
      otroJugador.obtener_capital <=> obtener_capital
    end
    
    def comprar(titulo)
      
    end
    
    def construir_casa(ip)
      
    end
    
    def construir_hotel(ip)
      
    end
    
    def en_bancarrota()
      @saldo <= 0
    end
    
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado())
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        @diario.ocurre_evento("Se encarcela al jugador:  " + @nombre)
      end
      @encarcelado
    end
    
    def existe_la_propiedad(ip)
      @propiedades.include?(@propiedades.at(ip))
    end
    
    def hipotecar(ip)
      
    end
    
    def modificar_saldo(cantidad)
      @saldo = @saldo + cantidad
      @diario.ocurre_evento("Se ha modificado el saldo en: " + cantidad);
      true
    end
    
    def mover_a_casilla(num_casilla)
      salida = false
      if !@encarcelado
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        @diario.ocurre_evento("Jugador movido a la casilla:  #{@nombre}")
        salida = true
      end
      salida
    end
    
    def obtener_salvoconducto(sorpresa)
      salida = false
      if !@encarcelado
        @salvoconducto = sorpresa
        salida = true
      end
      salida
    end
    
    def paga(cantidad)
      modificar_saldo(cantidad*-1)
    end
    
    def paga_alquiler(cantidad)
      salida = false
      if !@encarcelado
        salida = paga(cantidad)
      end
      salida
    end
    
    def paga_impuesto(cantidad)
      salida = false
      if !@encarcelado
        salida = paga(cantidad)
      end
      salida
    end
    
    def pasa_por_salida()
      modificar_saldo(@@PASO_POR_SALIDA)
      @diario.ocurre_evento("Jugador pasa por la salida:  " + @nombre)
      true
    end
    
    def puede_comprar_casilla()
      if @encalcelado
        @puede_comprar = false
      else
        @puede_comprar = treu
      end
      @puede_comprar
    end
     
    
    def recibe(cantidad)
      salida = false
      if !@encarcelado
        salida = modificar_saldo(cantidad)
      end
      salida
    end
    
    def salir_carcel_pagando()
      salida = false
      if @encarcelado && puede_salir_carcel_pagando()
        paga(@@PRECIO_LIBERTAD)
        @encarcelado = false
        salida = true
        @diario.ocurre_evento("Jugador sale del la carcel")
      end
      salida
    end
    
    def salir_carcel_tirando()
      salida = @dado.salgo_de_la_carcel
      if salida
        @encarcelado = false
        @diario.ocurre_evento("Jugador sale del la carcel")
      end
      salida
    end
    
    def tiene_algo_que_gestionar()
      @propiedades.empty?
    end
    
    def tiene_salvoconducto()
      @salvoconducto != nil
    end
    
    def vender(ip)
      salida = false
      if !@encacelado && existe_la_propiedad(ip)
        salida = @propiedades.at(ip).vender(self)
        @propiedades.delete_at(ip)
        @diario.ocurre_evento("Jugador vende")
      end
      salida
    end
    
    def to_s
      to_s = "Nombre: #{@nombre}\nSaldo: #{@saldo}\nEncarcelado: #{@encarcelado}\n#{@num_casilla_actual}\n"
      
      if !@propiedades.empty?
        for p in @propiedades
          to_s += "#{p}"
        end
      end
      to_s
    end
    
  end
end

File: Prácticas/P2/Java/civitas/src/main/java/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Jose Padial
 */
public class Jugador implements Comparable<Jugador>{
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa salvoconducto;
    private Diario diario = Diario.getInstance();
    private Dado dado = Dado.getInstance();
    protected static int casasMax = 4;
    protected static int casasPorHotel = 4;
    protected boolean encarcelado;
    protected static int hotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    protected static float pasoPorSalida = 1000;
    protected static float precioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static float saldoInicial = 7500;
    
    Jugador(String nombre){
        this.nombre = nombre;
        this.encarcelado = false;
        this.propiedades = new ArrayList();
        this.saldo = saldoInicial;
        this.puedeComprar = false;
    }
    
    protected Jugador(Jugador otro){
        this.encarcelado = otro.isEncarcelado();
        this.nombre = otro.getNombre();
        this.saldo = otro.getSaldo();
        this.propiedades = otro.getPropiedades();
        this.numCasillaActual = otro.getNumCasillaActual();
        this.puedeComprar = otro.getPuedeComprar();
    }
    
    boolean cancelarHipoteca(int ip){
        return true;
    }
    
    int cantidadCasasHoteles(){
        int suma = 0;
        for(TituloPropiedad p : propiedades){
            suma += p.getNumCasas() + p.getNumHoteles();
        }
        return suma;
    }
    
    @Override
    public int compareTo(Jugador otro){
        int otroCapital = (int) otro.getSaldo();
        return (int) ((int) otroCapital - this.getSaldo());
    }
    
    boolean comprar(TituloPropiedad titulo){
       return true; 
    }
    
    boolean construirCasa(int ip){
        return true;
    }
    
    boolean construirHotel(int ip){
        return true;
    }
    
    protected boolean debeSerEncarcelado(){
        if(encarcelado == true)
            return false;
        else{
            if(!tieneSalvoconducto())
                return true;
            else{
                perderSalvoConducto();
                diario.ocurreEvento("Se libra de la carcel el jugador: " + nombre);
                return false;
            }
        }
    }
    
    boolean enBancarrota(){
        return saldo <= 0;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            diario.ocurreEvento("Se encarcela al jugador:  " + nombre);
        }
        return encarcelado;
    }
    
    boolean existeLaPropiedad(int ip){
       return propiedades.contains(propiedades.get(ip));
    }

    protected ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    private int getCasasMax() {
        return casasMax;
    }

    int getCasasPorHotel() {
        return casasPorHotel;
    }

    public boolean isEncarcelado() {
        return encarcelado;
    }

    private int getHotelesMax() {
        return hotelesMax;
    }

    protected String getNombre() {
        return nombre;
    }

    int getNumCasillaActual() {
        return numCasillaActual;
    }

    private float getPrecioLibertad() {
        return precioLibertad;
    }

    boolean getPuedeComprar() {
        return puedeComprar;
    }

    protected float getSaldo() {
        return saldo;
    }
    
    private float getPremioPasoSalida(){
        return pasoPorSalida;
    }
    
    boolean hipotecar(int ip){
        return true;
    }
    
    boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        diario.ocurreEvento("Se ha modificado el saldo en: " + cantidad);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if(encarcelado == true)
            return false;
        else{
            numCasillaActual = numCasilla;
            puedeComprar = false;
            diario.ocurreEvento("Jugador movido a la casilla: " + numCasilla);
            return true;
        }
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        if(encarcelado == true)
            return false;
        else{
            salvoconducto = sorpresa;
            return true;
        }
    }
    
    boolean paga(float cantidad){
        return modificarSaldo(cantidad*-1);
    }
    
    boolean pagaAlquiler(float cantidad){
        if(encarcelado == true)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean pagaImpuesto(float cantidad){
        if(encarcelado == true)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean pasaPorSalida(){
        modificarSaldo(pasoPorSalida);
        diario.ocurreEvento("Jugador: " + nombre + " pasa por salida");
        return true;
    }
    
    private void perderSalvoConducto(){
        salvoconducto.usadas();
        salvoconducto = null;
    }
    
    boolean puedeComprarCasilla(){
        if(encarcelado == true)
            puedeComprar = false;
        else
            puedeComprar = true;
        return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        return saldo >= precioLibertad;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        if(puedoGastar(propiedad.getPrecioEdificar()) && propiedad.getNumCasas() < 4)
            return true;
        return false;
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        if(puedoGastar(propiedad.getPrecioEdificar()) && propiedad.getNumHoteles()< 4 && propiedad.getNumCasas() == 4)
            return true;
        return false;
    }
    
    private boolean puedoGastar(float precio){
        if(encarcelado == true)
            return false;
        else
            return saldo >= precio;
    }
    
    boolean recibe(float cantidad){
        if(encarcelado == true)
            return false;
        else
            return modificarSaldo(cantidad);
    }
    
    boolean salirCarcelPagando(){
        if(encarcelado == true && puedeSalirCarcelPagando()){
            paga(precioLibertad);
            encarcelado = false;
            diario.ocurreEvento("Jugador: " + nombre + " sale carcel pagando");
            return true;
        }
        return false;
    }
    
    boolean salirCarcelTirando(){
        if(encarcelado == true){
            if(dado.salgoDeLaCarcel()){
                encarcelado = false;
                diario.ocurreEvento("Jugador: " + nombre + " sale carcel tirando");
                return true;
            }
        }
        return false;
    }
    
    boolean tieneAlgoQueGestionar(){
        return !propiedades.isEmpty();
    }
    
    boolean tieneSalvoconducto(){
        return salvoconducto != null;
    }
    
    boolean vender(int ip){
        if(encarcelado == true)
            return false;
        else{
            if(existeLaPropiedad(ip)){
                propiedades.get(ip).vender(this);
                propiedades.remove(ip);
                diario.ocurreEvento("Propiedad vendida por jugador: " + nombre);
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "Jugador{" + "propiedades=" + propiedades + ", salvoconducto=" + salvoconducto + ", diario=" + diario.leerEvento() + ", encarcelado=" + encarcelado + ", nombre=" + nombre + ", numCasillaActual=" + numCasillaActual + ", puedeComprar=" + puedeComprar + ", saldo=" + saldo + '}';
    }
}
